File: src/electrickery/conversion/CurrentSource.java
/**
 * <copyright>
 * </copyright>
 *
 * $Id$
 */
package electrickery.conversion;

import org.eclipse.emf.common.util.EList;


/**
 * <!-- begin-user-doc -->
 * A representation of the model object '<em><b>Current Source</b></em>'.
 * <!-- end-user-doc -->
 *
 * <!-- begin-model-doc -->
 * Ideal current source.
 * 
 * ISource maintains a positive sequence for harmonic scans.  If you want zero
 * sequence, use three single-phase ISource.
 * <!-- end-model-doc -->
 *
 * <p>
 * The following features are supported:
 * <ul>
 *   <li>{@link electrickery.conversion.CurrentSource#getBus1 <em>Bus1</em>}</li>
 *   <li>{@link electrickery.conversion.CurrentSource#getAmps <em>Amps</em>}</li>
 *   <li>{@link electrickery.conversion.CurrentSource#getAngle <em>Angle</em>}</li>
 *   <li>{@link electrickery.conversion.CurrentSource#getFrequency <em>Frequency</em>}</li>
 *   <li>{@link electrickery.conversion.CurrentSource#getPhases <em>Phases</em>}</li>
 *   <li>{@link electrickery.conversion.CurrentSource#getScanType <em>Scan Type</em>}</li>
 * </ul>
 * </p>
 *
 * @see electrickery.conversion.ConversionPackage#getCurrentSource()
 * @model
 * @generated
 */
public interface CurrentSource extends PowerConversionElement {
	/**
	 * Returns the value of the '<em><b>Bus1</b></em>' attribute.
	 * <!-- begin-user-doc -->
	 * <!-- end-user-doc -->
	 * <!-- begin-model-doc -->
	 * Name of bus to which source is connected.
	 * <!-- end-model-doc -->
	 * @return the value of the '<em>Bus1</em>' attribute.
	 * @see #setBus1(String)
	 * @see electrickery.conversion.ConversionPackage#getCurrentSource_Bus1()
	 * @model
	 * @generated
	 */
	String getBus1();

	/**
	 * Sets the value of the '{@link electrickery.conversion.CurrentSource#getBus1 <em>Bus1</em>}' attribute.
	 * <!-- begin-user-doc -->
	 * <!-- end-user-doc -->
	 * @param value the new value of the '<em>Bus1</em>' attribute.
	 * @see #getBus1()
	 * @generated
	 */
	void setBus1(String value);

	/**
	 * Returns the value of the '<em><b>Amps</b></em>' attribute.
	 * <!-- begin-user-doc -->
	 * <!-- end-user-doc -->
	 * <!-- begin-model-doc -->
	 * Magnitude of current source, each phase, in Amps.
	 * <!-- end-model-doc -->
	 * @return the value of the '<em>Amps</em>' attribute.
	 * @see #setAmps(double)
	 * @see electrickery.conversion.ConversionPackage#getCurrentSource_Amps()
	 * @model
	 * @generated
	 */
	double getAmps();

	/**
	 * Sets the value of the '{@link electrickery.conversion.CurrentSource#getAmps <em>Amps</em>}' attribute.
	 * <!-- begin-user-doc -->
	 * <!-- end-user-doc -->
	 * @param value the new value of the '<em>Amps</em>' attribute.
	 * @see #getAmps()
	 * @generated
	 */
	void setAmps(double value);

	/**
	 * Returns the value of the '<em><b>Angle</b></em>' attribute.
	 * <!-- begin-user-doc -->
	 * <!-- end-user-doc -->
	 * <!-- begin-model-doc -->
	 * Phase angle in degrees of first phase. Phase shift between phases is
	 * assumed 120 degrees when number of phases <= 3.
	 * <!-- end-model-doc -->
	 * @return the value of the '<em>Angle</em>' attribute.
	 * @see #setAngle(double)
	 * @see electrickery.conversion.ConversionPackage#getCurrentSource_Angle()
	 * @model
	 * @generated
	 */
	double getAngle();

	/**
	 * Sets the value of the '{@link electrickery.conversion.CurrentSource#getAngle <em>Angle</em>}' attribute.
	 * <!-- begin-user-doc -->
	 * <!-- end-user-doc -->
	 * @param value the new value of the '<em>Angle</em>' attribute.
	 * @see #getAngle()
	 * @generated
	 */
	void setAngle(double value);

	/**
	 * Returns the value of the '<em><b>Frequency</b></em>' attribute.
	 * The default value is <code>"60.0"</code>.
	 * <!-- begin-user-doc -->
	 * <!-- end-user-doc -->
	 * <!-- begin-model-doc -->
	 * Source frequency.  Defaults to circuit fundamental frequency.
	 * <!-- end-model-doc -->
	 * @return the value of the '<em>Frequency</em>' attribute.
	 * @see #setFrequency(double)
	 * @see electrickery.conversion.ConversionPackage#getCurrentSource_Frequency()
	 * @model default="60.0"
	 * @generated
	 */
	double getFrequency();

	/**
	 * Sets the value of the '{@link electrickery.conversion.CurrentSource#getFrequency <em>Frequency</em>}' attribute.
	 * <!-- begin-user-doc -->
	 * <!-- end-user-doc -->
	 * @param value the new value of the '<em>Frequency</em>' attribute.
	 * @see #getFrequency()
	 * @generated
	 */
	void setFrequency(double value);

	/**
	 * Returns the value of the '<em><b>Phases</b></em>' attribute.
	 * <!-- begin-user-doc -->
	 * <!-- end-user-doc -->
	 * <!-- begin-model-doc -->
	 * Number of phases. For 3 or less, phase shift is 120 degrees.
	 * <!-- end-model-doc -->
	 * @return the value of the '<em>Phases</em>' attribute.
	 * @see #setPhases(int)
	 * @see electrickery.conversion.ConversionPackage#getCurrentSource_Phases()
	 * @model
	 * @generated
	 */
	int getPhases();

	/**
	 * Sets the value of the '{@link electrickery.conversion.CurrentSource#getPhases <em>Phases</em>}' attribute.
	 * <!-- begin-user-doc -->
	 * <!-- end-user-doc -->
	 * @param value the new value of the '<em>Phases</em>' attribute.
	 * @see #getPhases()
	 * @generated
	 */
	void setPhases(int value);

	/**
	 * Returns the value of the '<em><b>Scan Type</b></em>' attribute.
	 * The literals are from the enumeration {@link electrickery.conversion.sequenceType}.
	 * <!-- begin-user-doc -->
	 * <!-- end-user-doc -->
	 * <!-- begin-model-doc -->
	 * Maintain specified sequence for harmonic solution.  Otherwise, angle
	 * between phases rotates with harmonic.
	 * <!-- end-model-doc -->
	 * @return the value of the '<em>Scan Type</em>' attribute.
	 * @see electrickery.conversion.sequenceType
	 * @see #setScanType(sequenceType)
	 * @see electrickery.conversion.ConversionPackage#getCurrentSource_ScanType()
	 * @model
	 * @generated
	 */
	sequenceType getScanType();

	/**
	 * Sets the value of the '{@link electrickery.conversion.CurrentSource#getScanType <em>Scan Type</em>}' attribute.
	 * <!-- begin-user-doc -->
	 * <!-- end-user-doc -->
	 * @param value the new value of the '<em>Scan Type</em>' attribute.
	 * @see electrickery.conversion.sequenceType
	 * @see #getScanType()
	 * @generated
	 */
	void setScanType(sequenceType value);

	/**
	 * <!-- begin-user-doc -->
	 * <!-- end-user-doc -->
	 * @model kind="operation" lower="2" upper="2"
	 * @generated
	 */
	EList<Double> getBaseCurrent();

} // CurrentSource

File: dss/conversion/CurrentSource.py
# Copyright (C) 2010 Richard Lincoln
#
# This library is free software; you can redistribute it and/or
# modify it under the terms of the GNU Lesser General Public
# License as published by the Free Software Foundation; either
# version 2.1 of the License, or (at your option) any later version.
#
# This library is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
# Lesser General Public License for more details.
#
# You should have received a copy of the GNU Lesser General Public
# License along with this library; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA, USA

from scipy.sparse import csr_matrix as sparse

from dss.conversion.PowerConversionElement import PowerConversionElement

class CurrentSource(PowerConversionElement):
    """Ideal current source.  ISource maintains a positive sequence for
    harmonic scans.  If you want zero sequence, use three single-phase ISource.
    """

    def __init__(self, bus1='', amps=0.0, angle=0.0, frequency=60.0, phases=0,
            scanType="Positive", *args, **kw_args):
        """Initialises a new 'CurrentSource' instance.
        """
        #: Name of bus to which source is connected.
        self.bus1 = bus1

        #: Magnitude of current source, each phase, in Amps.
        self.amps = amps

        #: Phase angle in degrees of first phase. Phase shift between phases
        #  is assumed 120 degrees when number of phases <= 3.
        self.angle = angle

        #: Source frequency.  Defaults to circuit fundamental frequency.
        self.frequency = frequency

        #: Number of phases. For 3 or less, phase shift is 120 degrees.
        self.phases = phases

        #: Maintain specified sequence for harmonic solution.  Otherwise,
        #  angle between phases rotates with harmonic. Values are: "Positive",
        #  "Zero", "None"
        self.scanType = scanType

        #: Injection current.
        inj_current = 0.0

        super(CurrentSource, self).__init__(*args, **kw_args)


    @property
    def nPhases(self):
        """Number of phases. For 3 or less, phase shift is 120 degrees.
        Valid 'phase' values are: "BCN", "ACN", "AB", "A", "B", "ABCN",
        "AC", "N", "AN", "C", "ABN", "BN", "ABC", "BC", "CN"
        """
        phases = self.phases
        if phases in ["ABCN", "ABC"]:
            n_phases = 3
        elif phases in ["BCN", "ACN", "ABN", "AB", "BC"]:
            n_phases = 2
        elif phases in ["AN", "BN", "CN", "A", "B", "C"]:
            n_phases = 1
        else:
            n_phases = 0

        return n_phases


    def calcYPrim(self):
        """Returns the primitive Y (admittance) matrix for this element alone.
        """
        n_phases = self.nPhases
        y_order = n_phases * len(self.terminals)

        self.y_prim_series = sparse((y_order, y_order))
        self.y_prim        = sparse((y_order, y_order))

        # Yprim = 0 for Ideal Current Source; just leave it zeroed
        # TODO: Zero out rows and columns for open conductors.

        self.y_prim_invalid = False
